realsense2_camera/include/mavros_timestamp.h

```diff
@@ -0,0 +1,432 @@
+      // complete information if needed
+      cache_queue_type& cache_entry = cache_queue_[channel][trigger_seq];
+      cache_entry.trigger_info = true;
+      cache_entry.stamp_trigger = trigger_time;
+    }
+  }
+
+  /*
+   * Checks all cache entries in cache_queue_ and publishes those that
+   *  - have both camera info and trigger info filled out
+   *  - have sequence_id = last_published_trigger_seq_ + 1 ( so we only publish in order)
+   *  Runs as long as messages are available that satisfy this.
+   */
+  void releaseCacheEntries(const t_channel_id& channel){
+    bool entry_released = false;
+
+    // release all entries that are complete.
+    do {
+      uint32_t seq_trigger = last_published_trigger_seq_[channel] + 1;
+
+      // check if "next" trigger sequence exists and is ready to be released
+      auto it_next_sequence = cache_queue_[channel].find(seq_trigger);
+
+      // if item exists and has all necessary info, release it
+      entry_released = (it_next_sequence != cache_queue_[channel].end()) &&
+          (it_next_sequence->second.camera_info && it_next_sequence->second.trigger_info);
+
+      if(entry_released) {
+        auto entry = it_next_sequence->second;
+
+        // adjust timestamp
+        ros::Time midpoint_exposure = shiftTimestampToMidExposure(entry.stamp_trigger,
+                                                                   entry.exposure);
+
+        // callback to publish if there is a cached frame
+        //  - sometimes we add camera sequences without cached frames, e.g. if there are no subscribers
+        //  - e.g. by calling addCameraSequence
+        if(entry.frame != nullptr) {
+          callback_(channel, midpoint_exposure, entry.frame);
+        }
+        // output delay message and log
+        last_delay_ = midpoint_exposure.toSec() - entry.stamp_camera.toSec();
+
+        geometry_msgs::PointStamped msg;
+        msg.header.stamp = midpoint_exposure;
+        msg.point.x = last_delay_;
+        msg.point.y = entry.seq_camera;
+        msg.point.z = seq_trigger;
+        delay_pub_.publish(msg);
+
+        ROS_DEBUG_STREAM(kLogPrefix << "|cache|= " << cache_queue_[channel].size() - 1 << " Frame w seq_c=" << entry.seq_camera << ", seq_t=" << seq_trigger << " released w t_diff=" << last_delay_);
+
+        // cleanup
+        last_published_trigger_seq_[channel]++;
+        cache_queue_[channel].erase(it_next_sequence);
+
+      } else {
+        ROS_DEBUG_STREAM(kLogPrefix << "|cache|= " << cache_queue_[channel].size() << ". ");
+      }
+
+    } while(entry_released);
+
+    // cleanup old entries - warning - special structure of for-loop needed to traverse while removing.
+    for(auto it = cache_queue_[channel].cbegin(); it != cache_queue_[channel].cend();/*no inc++*/){
+      if(static_cast<int32_t>(it->first) < last_published_trigger_seq_[channel]){
+        // remove all entries that cannot be published anymore
+        cache_queue_[channel].erase(it++);
```

Comment:
I had no idea you could use a const_iterator for erasing

Reply:
:rocket: